### DISCUSSION TITLE
HOWTO: Access credentials within custom tools / custom JS functions?

### DISCUSSION THREAD

**Discussion by dkindlund** [FlowiseAI/Flowise#1548] [Q&A]
For example, I'd like to write a custom tool that accesses a Redis server where I already have the credentials stored natively in Flowise.  Is there any way I can easily access those credentials in the context of a custom tool or custom JS function?

**Comment by HenryHengZJ** [comment]
you'll have to use `variables` for that

**Reply by dkindlund** [reply]
Hey @HenryHengZJ , I think you're talking about the `variables` field in the `INodeParams` interface -- but I'm trying to figure out how to access that field *within* the Custom JavaScript function.  If you have some sample code that shows how to do this, I'd appreciate it.

**Reply by HenryHengZJ** [reply]
just merged the PR that enables using `$vars` in custom function - https://github.com/FlowiseAI/Flowise/pull/1558. That way you can create your credentials as `vars`, although its not ideal, but its a workaround for now, until we bring in the ability to take `$credential`

**Reply by dkindlund** [reply]
Got it; thanks for the clarification, @HenryHengZJ !

**Reply by j-d-carmichael** [reply]
@HenryHengZJ is pulling in the $credential object available yet in custom tools?

**Reply by bharadwaj-ramachandran-tfs** [reply]
hi @HenryHengZJ  is this still feature available in enterprise?

**Reply by HenryHengZJ** [reply]
Unfortunately not, we recommend to use runtime Variables (https://docs.flowiseai.com/using-flowise/variables#runtime) which you can then specify the value in `.env` file
